### PR TITLE
ci: generate ts typings for grpc client stub

### PR DESCRIPTION
**Pull request by sebastiendan**
# Description

#229 introduced JavaScript gRPC client stubs generation through a new CI workflow. This PR adds the generation of TypeScript typings so that TypeScript clients can better use our client stubs.

Additionally, this PR changes the npm package publication flow to the following one:
- First of all, the CI workflow will only run when
  - the push on `main` contains changes to protobuf files
  - the PR contains changes to protobuf files (coming from the active update or previous ones)
- On PR updates, a step will check whether the npm package version was bumped and make the job fail if not
- On merge on `main`, the `npm publish` step is run if the above was successful

Testing this new logic:
- [run](https://github.com/topos-network/topos/actions/runs/5072802220) that shows a failure because of missing version bump
- [run](https://github.com/topos-network/topos/actions/runs/5072886924/jobs/9111190112#step:9:20) that shows a success because version was bumped

## Additions and Changes

- Add the `grpc_tools_node_protoc_ts` npm package as dep
- Install `protoc` in the CI workflow
- Use `protoc` with a bin packaged in `grpc_tools_node_protoc_ts` to generate typings
- Bump npm package version to `0.1.1`

## PR Checklist:

- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: .github/workflows/grpc-client-stubs.yml [hunk 1]
```diff
@@ -38,6 +45,20 @@ jobs:
...
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/shared/v1/frost.proto &&
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/shared/v1/stark_proof.proto &&
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/shared/v1/subnet.proto &&
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/shared/v1/uuid.proto &&
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/uci/v1/certification.proto &&
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/tce/v1/api.proto &&
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/tce/v1/console.proto &&
+          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated topos/tce/v1/synchronization.proto
```
Comment (Freyskeyd) [inline .github/workflows/grpc-client-stubs.yml:60]: You can use something like that:

```suggestion
        run: >
          protoc --plugin=protoc-gen-ts=./node_modules/.bin/protoc-gen-ts --ts_out=grpc_js:./generated -I . $(find ./topos -name '*.proto')
```

**Review by Freyskeyd — APPROVED** [review]
Could you rename the github action workflow to `gRPC client stubs generation` as we could add more language to this workflow at some point.

**Review by gruberb — APPROVED** [review]
LGTM 👍

**Review by gruberb — COMMENTED** [review]
Looks good, thanks for the detailed write-up! 

Just wondering if there is a scenario where the types of the JS gRPC client could be newer than the current release? Merging with `main` doesn't mean release, so the first logic seem to make more sense to me (because now we could have a situation with unreleased, but updated types in `main` the JS client is using, but the currently deployed `topos` network is not yet using).

**Review by gruberb — COMMENTED** [review]
Looks good, thanks for the detailed write-up! 

Just wondering if there is a scenario where the types of the JS gRPC client could be newer than the current release? Merging with `main` doesn't mean release, so the first logic seem to make more sense to me (because now we could have a situation with unreleased, but updated types in `main` the JS client is using, but the currently deployed `topos` network is not yet using).

**Comment by sebastiendan** [comment]
> Looks good, thanks for the detailed write-up!
> 
> Just wondering if there is a scenario where the types of the JS gRPC client could be newer than the current release? Merging with `main` doesn't mean release, so the first logic seem to make more sense to me (because now we could have a situation with unreleased, but updated types in `main` the JS client is using, but the currently deployed `topos` network is not yet using).

Housekeeping: This was replied on [Slack](https://toposware.slack.com/archives/C045B9XKRFE/p1685036136069189?thread_ts=1684953988.734629&cid=C045B9XKRFE)

**Pull request merged**